Add volume snapshot exercise (#3)
cloudscale-ch/cloud-exercises: exercises/02-storage.md

Can you grow the size of the volume without restarting the system?

Hint: Click on the question mark in the Cloud Control Panel near each volume for further instructions.

---

> ### `02.03` - Snapshot a Volume

Reviewer: For consistency with the rest:
```suggestion
### `02.03` - Snapshot A Volume
```

To be more precise:
```suggestion
### `02.03` - Snapshot A Root Volume
```

> Please be aware: Snapshots do not serve as a substitute for backups.

> Using the Cloud Control Panel, or your favourite tool, create a server with a root volume attached.

Reviewer: We tend to use american spelling, so that would be `favorite`.

> Using the Cloud Control Panel, or your favourite tool, create a server with a root volume attached.

Reviewer: > create a server with a root volume attached.

Since you cannot start a server without root volume, I would just write: Create a server.

Create a snapshot of your root volume, using our [API](https://www.cloudscale.ch/en/api/v1#volume-snapshots).

> Once the operation completes with success, start your maintenance tasks. As of this excercice, you can delete an important file, such as `/etc/passwd` to simulate a failed maintenance. [Revert the previously created snapshot](https://www.cloudscale.ch/en/api/v1#revert-volume-to-snapshot) to restore the system to the last running state.

Reviewer: ```suggestion
Once the operation completes successfully, start your maintenance tasks. For this exercise, you can delete an important file, such as `/etc/passwd`, to simulate a failed maintenance. [Revert the previously created snapshot](https://www.cloudscale.ch/en/api/v1#revert-volume-to-snapshot) to restore the system to the last working state.
```

> Does the file `/etc/passwd` exist again?

Reviewer: ```suggestion
Once the VM has booted, does `/etc/passwd` reappear?
```

Can you grow the size of the volume without restarting the system?

Hint: Click on the question mark in the Cloud Control Panel near each volume for further instructions.

---

### `02.03` - Snapshot A Root Volume

> Please be aware: Snapshots do not serve as a substitute for backups.

Using the Cloud Control Panel, or your favorite tool, create a new server.

Create a snapshot of your root volume, using our [API](https://www.cloudscale.ch/en/api/v1#volume-snapshots).

> > A snapshot captures the state of the disk at the moment it is taken. However, only changes that are fully written and flushed to disk before the snapshot will be included in it. In contrast, any writes that were in progress at the time of the snapshot may not be fully backed up.

Reviewer: To avoid the use of `backup`:
```suggestion
> A snapshot captures the state of the disk at the moment it is taken. However, only changes that are fully written and flushed to disk before the snapshot will be included in it. In contrast, any writes that were in progress at the time of the snapshot may not be fully captured.
```

Once the operation completes successfully, start your maintenance tasks. For this exercise, you can delete an important file, such as `/etc/passwd`, to simulate a failed maintenance. [Revert the previously created snapshot](https://www.cloudscale.ch/en/api/v1#revert-volume-to-snapshot) to restore the system to the last working state.

Once the VM has booted, does `/etc/passwd` reappear?

---

### `02.04` - Snapshot A (Non-)Root Volume

> Please be aware: Snapshots do not serve as a substitute for backups.

Using the Cloud Control Panel, or your favorite tool, create a new server and add an additional non-root volume to it.
Format the volume to ext4 with `sudo mkfs.ext4 /dev/sdb` and mount it to `/mnt`.

Create a new file on that volume and ensure it is synced to disk with fsync: `sudo dd if=/dev/zero of=/mnt/included count=1 bs=1M conv=fsync`.

> > A snapshot captures the state of the disk at the moment it is taken. However, only changes that are fully written and flushed to disk before the snapshot will be included in it. In contrast, any writes that were in progress at the time of the snapshot may not be fully backed up.

Reviewer: See above.

Create a snapshot of the additional volume, using our [API](https://www.cloudscale.ch/en/api/v1#volume-snapshots).

Now, create a second file that is not included in the snapshot: `sudo dd if=/dev/zero of=/mnt/not-included count=1 bs=1M`.

[Revert the previously created snapshot](https://www.cloudscale.ch/en/api/v1#revert-volume-to-snapshot) to restore the volume to its previous state.
You will encounter the following error message unless you first shut down your server and detach the volume before executing the restore command:
```
{
  "detail": "Cannot revert non-root volumes while they are attached to a server."
}
```
To successfully perform the revert operation, ensure that the server is powered down and the volume is detached before retrying.

> 

Reviewer: Nit: remove this extra line.
